### PR TITLE
Optimize Ascii.Equals for small values

### PR DESCRIPTION
**Pull request by yesmey**
An attempt to optimize Ascii.Equals for smaller values

Changes include:
- 64bit register simd
- Use better widening for 128bit vector (same as the 256/512 path)
- Mask the ascii check into the same comparison for 256/512
- Changed Sse2.UnpackLow to Vector128.WidenLower and removed an unused Vector64 path (Vector128.WidenLower will use vpmovzxbw when AVX2 is available)

Now that the code is shared across the different vector sizes now, it should be easy to convert to ISimdVector :) but it doesn't have any widening methods yet.

<details>
<summary>Benchmarks:</summary>

```

BenchmarkDotNet v0.13.10-nightly.20231019.90, Windows 11 (10.0.22631.2506/23H2/2023Update/SunValley3)
AMD Ryzen 7 7800X3D, 1 CPU, 16 logical and 8 physical cores
.NET SDK 9.0.100-alpha.1.23553.1
  [Host]     : .NET 8.0.0 (8.0.23.47906), X64 RyuJIT AVX2
  Job-DWACHG : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-UKFQFX : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2

PowerPlanMode=00000000-0000-0000-0000-000000000000  Arguments=/p:EnableUnsafeBinaryFormatterSerialization=true  IterationTime=250.0000 ms  
LaunchCount=3  MaxIterationCount=20  MemoryRandomization=True  
MinIterationCount=15  WarmupCount=1  

```
| Method                           | Job        | Toolchain                                                                                                        | Size | Mean     | Error     | StdDev    | Median   | Min      | Max      | Ratio | RatioSD | Allocated | Alloc Ratio |
|--------------------------------- |----------- |----------------------------------------------------------------------------------------------------------------- |----- |---------:|----------:|----------:|---------:|---------:|---------:|------:|--------:|----------:|------------:|
| Equals_Bytes                     | Job-DWACHG | main        | 6    | 2.807 ns | 0.0136 ns | 0.0258 ns | 2.801 ns | 2.767 ns | 2.879 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Bytes                     | Job-UKFQFX | PR | 6    | 2.857 ns | 0.0152 ns | 0.0289 ns | 2.848 ns | 2.817 ns | 2.945 ns |  1.02 |    0.01 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_Chars                     | Job-DWACHG | main        | 6    | 3.426 ns | 0.1181 ns | 0.2642 ns | 3.334 ns | 3.014 ns | 4.005 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Chars                     | Job-UKFQFX | PR | 6    | 1.926 ns | 0.0083 ns | 0.0157 ns | 1.920 ns | 1.913 ns | 2.005 ns |  0.56 |    0.04 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_Bytes_Chars               | Job-DWACHG | main        | 6    | 3.456 ns | 0.0905 ns | 0.2025 ns | 3.389 ns | 3.200 ns | 4.054 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Bytes_Chars               | Job-UKFQFX | PR | 6    | 2.914 ns | 0.0304 ns | 0.0578 ns | 2.896 ns | 2.848 ns | 3.104 ns |  0.84 |    0.05 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Bytes       | Job-DWACHG | main        | 6    | 1.330 ns | 0.0036 ns | 0.0069 ns | 1.329 ns | 1.322 ns | 1.348 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Bytes       | Job-UKFQFX | PR | 6    | 1.330 ns | 0.0038 ns | 0.0073 ns | 1.328 ns | 1.322 ns | 1.346 ns |  1.00 |    0.01 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Chars       | Job-DWACHG | main        | 6    | 1.333 ns | 0.0048 ns | 0.0091 ns | 1.327 ns | 1.323 ns | 1.353 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Chars       | Job-UKFQFX | PR | 6    | 1.333 ns | 0.0043 ns | 0.0081 ns | 1.330 ns | 1.324 ns | 1.348 ns |  1.00 |    0.01 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Bytes_Chars | Job-DWACHG | main        | 6    | 1.330 ns | 0.0035 ns | 0.0066 ns | 1.326 ns | 1.324 ns | 1.344 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Bytes_Chars | Job-UKFQFX | PR | 6    | 1.327 ns | 0.0031 ns | 0.0059 ns | 1.325 ns | 1.321 ns | 1.346 ns |  1.00 |    0.01 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_Bytes                     | Job-DWACHG | main        | 9    | 4.017 ns | 0.0563 ns | 0.1199 ns | 3.987 ns | 3.863 ns | 4.306 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Bytes                     | Job-UKFQFX | PR | 9    | 1.930 ns | 0.0054 ns | 0.0103 ns | 1.928 ns | 1.917 ns | 1.948 ns |  0.48 |    0.01 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_Chars                     | Job-DWACHG | main        | 9    | 2.121 ns | 0.0045 ns | 0.0086 ns | 2.118 ns | 2.111 ns | 2.144 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Chars                     | Job-UKFQFX | PR | 9    | 2.324 ns | 0.0054 ns | 0.0103 ns | 2.321 ns | 2.313 ns | 2.347 ns |  1.10 |    0.01 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_Bytes_Chars               | Job-DWACHG | main        | 9    | 4.408 ns | 0.0069 ns | 0.0132 ns | 4.404 ns | 4.391 ns | 4.436 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Bytes_Chars               | Job-UKFQFX | PR | 9    | 2.135 ns | 0.0073 ns | 0.0139 ns | 2.131 ns | 2.118 ns | 2.198 ns |  0.48 |    0.00 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Bytes       | Job-DWACHG | main        | 9    | 1.326 ns | 0.0033 ns | 0.0062 ns | 1.324 ns | 1.319 ns | 1.344 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Bytes       | Job-UKFQFX | PR | 9    | 1.333 ns | 0.0041 ns | 0.0077 ns | 1.332 ns | 1.323 ns | 1.350 ns |  1.01 |    0.01 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Chars       | Job-DWACHG | main        | 9    | 1.528 ns | 0.0040 ns | 0.0075 ns | 1.525 ns | 1.521 ns | 1.550 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Chars       | Job-UKFQFX | PR | 9    | 1.533 ns | 0.0069 ns | 0.0132 ns | 1.529 ns | 1.519 ns | 1.586 ns |  1.00 |    0.01 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Bytes_Chars | Job-DWACHG | main        | 9    | 1.339 ns | 0.0064 ns | 0.0121 ns | 1.340 ns | 1.321 ns | 1.360 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Bytes_Chars | Job-UKFQFX | PR | 9    | 1.560 ns | 0.0140 ns | 0.0267 ns | 1.562 ns | 1.521 ns | 1.603 ns |  1.16 |    0.02 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_Bytes                     | Job-DWACHG | main        | 15   | 5.272 ns | 0.0369 ns | 0.0703 ns | 5.238 ns | 5.204 ns | 5.463 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Bytes                     | Job-UKFQFX | PR | 15   | 1.927 ns | 0.0034 ns | 0.0064 ns | 1.925 ns | 1.919 ns | 1.944 ns |  0.37 |    0.00 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_Chars                     | Job-DWACHG | main        | 15   | 2.139 ns | 0.0276 ns | 0.0558 ns | 2.125 ns | 2.117 ns | 2.501 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Chars                     | Job-UKFQFX | PR | 15   | 2.318 ns | 0.0044 ns | 0.0083 ns | 2.315 ns | 2.310 ns | 2.345 ns |  1.08 |    0.03 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_Bytes_Chars               | Job-DWACHG | main        | 15   | 6.786 ns | 0.0070 ns | 0.0132 ns | 6.780 ns | 6.773 ns | 6.826 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Bytes_Chars               | Job-UKFQFX | PR | 15   | 2.145 ns | 0.0072 ns | 0.0137 ns | 2.139 ns | 2.128 ns | 2.177 ns |  0.32 |    0.00 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Bytes       | Job-DWACHG | main        | 15   | 1.329 ns | 0.0031 ns | 0.0059 ns | 1.327 ns | 1.324 ns | 1.345 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Bytes       | Job-UKFQFX | PR | 15   | 1.331 ns | 0.0038 ns | 0.0071 ns | 1.329 ns | 1.323 ns | 1.349 ns |  1.00 |    0.01 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Chars       | Job-DWACHG | main        | 15   | 1.654 ns | 0.0809 ns | 0.1635 ns | 1.542 ns | 1.522 ns | 2.009 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Chars       | Job-UKFQFX | PR | 15   | 1.533 ns | 0.0080 ns | 0.0151 ns | 1.527 ns | 1.520 ns | 1.603 ns |  0.93 |    0.09 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Bytes_Chars | Job-DWACHG | main        | 15   | 1.347 ns | 0.0092 ns | 0.0175 ns | 1.345 ns | 1.324 ns | 1.380 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Bytes_Chars | Job-UKFQFX | PR | 15   | 1.548 ns | 0.0098 ns | 0.0186 ns | 1.538 ns | 1.525 ns | 1.606 ns |  1.15 |    0.02 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_Bytes                     | Job-DWACHG | main        | 17   | 2.122 ns | 0.0054 ns | 0.0102 ns | 2.120 ns | 2.108 ns | 2.148 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Bytes                     | Job-UKFQFX | PR | 17   | 2.322 ns | 0.0058 ns | 0.0110 ns | 2.317 ns | 2.310 ns | 2.349 ns |  1.09 |    0.01 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_Chars                     | Job-DWACHG | main        | 17   | 2.192 ns | 0.0500 ns | 0.0952 ns | 2.127 ns | 2.117 ns | 2.345 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Chars                     | Job-UKFQFX | PR | 17   | 2.323 ns | 0.0054 ns | 0.0102 ns | 2.319 ns | 2.313 ns | 2.348 ns |  1.06 |    0.04 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_Bytes_Chars               | Job-DWACHG | main        | 17   | 2.715 ns | 0.0056 ns | 0.0107 ns | 2.712 ns | 2.703 ns | 2.743 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Bytes_Chars               | Job-UKFQFX | PR | 17   | 2.327 ns | 0.0080 ns | 0.0152 ns | 2.322 ns | 2.313 ns | 2.376 ns |  0.86 |    0.01 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Bytes       | Job-DWACHG | main        | 17   | 1.528 ns | 0.0040 ns | 0.0077 ns | 1.525 ns | 1.521 ns | 1.547 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Bytes       | Job-UKFQFX | PR | 17   | 1.530 ns | 0.0068 ns | 0.0130 ns | 1.525 ns | 1.521 ns | 1.591 ns |  1.00 |    0.01 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Chars       | Job-DWACHG | main        | 17   | 1.557 ns | 0.0093 ns | 0.0177 ns | 1.556 ns | 1.522 ns | 1.589 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Chars       | Job-UKFQFX | PR | 17   | 1.569 ns | 0.0127 ns | 0.0242 ns | 1.566 ns | 1.527 ns | 1.622 ns |  1.01 |    0.01 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Bytes_Chars | Job-DWACHG | main        | 17   | 1.729 ns | 0.0043 ns | 0.0082 ns | 1.727 ns | 1.720 ns | 1.746 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Bytes_Chars | Job-UKFQFX | PR | 17   | 1.732 ns | 0.0053 ns | 0.0100 ns | 1.729 ns | 1.720 ns | 1.761 ns |  1.00 |    0.01 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_Bytes                     | Job-DWACHG | main        | 128  | 2.514 ns | 0.1013 ns | 0.2159 ns | 2.422 ns | 2.281 ns | 2.968 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Bytes                     | Job-UKFQFX | PR | 128  | 2.502 ns | 0.0410 ns | 0.0917 ns | 2.476 ns | 2.378 ns | 2.730 ns |  1.00 |    0.08 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_Chars                     | Job-DWACHG | main        | 128  | 4.921 ns | 0.1212 ns | 0.2712 ns | 5.028 ns | 4.173 ns | 5.306 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Chars                     | Job-UKFQFX | PR | 128  | 6.005 ns | 0.2843 ns | 0.6359 ns | 6.013 ns | 4.854 ns | 9.322 ns |  1.22 |    0.14 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_Bytes_Chars               | Job-DWACHG | main        | 128  | 3.044 ns | 0.0704 ns | 0.1575 ns | 2.973 ns | 2.868 ns | 3.463 ns |  1.00 |    0.00 |         - |          NA |
| Equals_Bytes_Chars               | Job-UKFQFX | PR | 128  | 3.815 ns | 0.6698 ns | 1.4981 ns | 3.271 ns | 2.903 ns | 8.177 ns |  1.25 |    0.50 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Bytes       | Job-DWACHG | main        | 128  | 1.545 ns | 0.0173 ns | 0.0337 ns | 1.538 ns | 1.523 ns | 1.748 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Bytes       | Job-UKFQFX | PR | 128  | 1.535 ns | 0.0049 ns | 0.0094 ns | 1.534 ns | 1.524 ns | 1.557 ns |  0.99 |    0.02 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Chars       | Job-DWACHG | main        | 128  | 1.578 ns | 0.0125 ns | 0.0238 ns | 1.582 ns | 1.534 ns | 1.619 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Chars       | Job-UKFQFX | PR | 128  | 1.576 ns | 0.0120 ns | 0.0228 ns | 1.579 ns | 1.532 ns | 1.638 ns |  1.00 |    0.02 |         - |          NA |
|                                  |            |                                                                                                                  |      |          |           |           |          |          |          |       |         |           |             |
| Equals_DifferentCase_Bytes_Chars | Job-DWACHG | main        | 128  | 1.958 ns | 0.0162 ns | 0.0307 ns | 1.953 ns | 1.910 ns | 2.032 ns |  1.00 |    0.00 |         - |          NA |
| Equals_DifferentCase_Bytes_Chars | Job-UKFQFX | PR | 128  | 1.954 ns | 0.0267 ns | 0.0515 ns | 1.934 ns | 1.890 ns | 2.081 ns |  1.00 |    0.03 |         - |          NA |
															  |
</details>

Benchmark code: just added different sizes to System.Text.Perf_Ascii in dotnet/performance
```dotnet run -c Release -f net8.0 --filter System.Text.Perf_Ascii.Equals_* --memoryRandomization --launchCount 5 --corerun ... --artifacts ...```

### CODE REVIEW

**Comment by adamsitnik** [comment]
@gfoidl would you be interested in reviewing this PR?

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:115]: Why is this needed? It's the last elements, so that result can be returned (`true` / `false`) directly, w/o the need to jump the end of the method. Or does this result in better codegen? (Although I have to admin that with PGO it's now much harder for me to guess / expect the codegen here as the profile data can move the blocks around).

I'd revert these changes (same return as before) and let the JIT do it's thing.

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:179]: This could also just be `return TLoader.EqualAndAscii256(...)`?

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:480]: If the `PlainLoader` take the HW into account (see comment) above, this could become
```suggestion
            static virtual bool Use64Bit => IntPtr.Size == 8 && Vector128.IsHardwareAccelerated;
```
and remove the methods from the struct implementations.

Maybe a better name also? As `Use64Bit` doesn't take the HW into account.
Naming is hard, but something like `Is64BitAndHardwareAccelerated`?

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:618]:
```suggestion
                return ((vecWide[0] ^ right) | (vecWide[1] ^ rightNext)) == 0;
```

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:511]:
```suggestion
                return leftValues == rightValues && AllCharsInUInt64AreAscii<T>(leftValues));
```

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:525]:
```suggestion
                return leftValues == rightValues && AllCharsInVectorAreAscii(leftValues);
```
I'd clean-up the existing method exits in this type here too. Codegen should be the same as before (and I can't remember why it was written that way).

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:574]: I'm not against that assert, but do we need it?

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:597]: In the case that in L51 the Vec128.IsHardwareAccelarated = false this would blow-up here. There's really a test-hole?!

But when `ILoader.Use64Bit` (see above) take the HW into account it's safe.

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:656]: Plain `return` here (note: GH doesn't allow me to put on the suggestion)

**Review by gfoidl — CHANGES_REQUESTED** [review]
I think there's a bug in the case `Vector128.IsHardwareAccelerated = false` -- see comments for rationale.

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (yesmey) [reply src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:51]: I agree, I was thinking about doing that too. The 64bit without Vector128.IsHardwareAccelerated sounds like an uncommon scenario. Can look into why there aren't covered by tests when I have time the upcoming week

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (yesmey) [reply src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:115]: IIRC the JIT was adding extra return sections when I had it, but I can remove it for the simplicity

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (yesmey) [reply src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:574]: Not really, I will remove it

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (yesmey) [reply src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:525]: I thought so too, but this was actually intentional from the original author. A conditional return here is adding extra instructions.

The same code before and after:
```asm
kortestw  k1,k1
jne       near ptr M01_L11
vptest    ymm0,ymmword ptr [7FFC7E491580]
jne       near ptr M01_L11
```

```asm
kortestw  k1,k1
jae       near ptr M01_L00
vptest    ymm0,ymmword ptr [7FFC841B1600]
sete      r9b
movzx     r9d,r9b
test      r9d,r9d
je        near ptr M01_L00
```

I could add a comment here about it though. Any idea what issue is tracking this?

Comment (gfoidl) [reply src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:525]: > Any idea what issue is tracking this?

@EgorBo do you have any idea here? Can this be fixed in the JIT (as the natural way of writing the code produces more machine code).

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:525]: I think it's fine to prefer better code readability over such quirks as we constantly improve JIT over those

**Review by gfoidl — APPROVED** [review]
Assuming CI will pass --> LGTM

**Comment by yesmey** [comment]
Updated benchmarks. Note that the regressions comes mainly from different codegen when returning the bool condition vs true/false. See comment https://github.com/dotnet/runtime/pull/93191#discussion_r1382382151
Example of regression:
Before: https://godbolt.org/z/Gnsc84Gsf
After: https://godbolt.org/z/8dMhhPMTn

File: src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs
Comment (adamsitnik) [inline src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs:82]: this comment was valuable, please restore it

```suggestion
                        if (chr != '?') // ASCIIEncoding maps invalid ASCII to ?
```

File: src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs
Comment (adamsitnik) [inline src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs:46]: That is a lot of test cases, I am not sure how long it's going to take to run all of them with debug builds. 

Before your change, this test  was covering strings that were 0 to 127 chars long. Now the max length is reduced to 33. I am not convinced that this is the right thing to do, please revert this particular change or convince me that it's the right thing to do.

File: src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs
Comment (adamsitnik) [inline src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs:78]: Same as above, I don't see benefits of this change.

File: src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs
Comment (adamsitnik) [inline src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs:119]: I see the value of this change, but we should decrease the number of test cases. How about just focusing on Ascii letters? Because for other characters we would be duplicating other tests work.

**Review by adamsitnik — COMMENTED** [review]
The `Asii` changes LGTM, but please solve the merge conflict and address test comments before I hit the merge button.

Thank you for your contribution @yesmey !

@gfoidl big thanks for your review!

**Pull request closed without merging**
